### PR TITLE
Add ARRAY to test_all_types + IO and some clients

### LINKED ISSUE #10663: ARRAY and parquet: not quite complete
### What happens?

A table containing an ARRAY column won't be written to a parquet file.

### To Reproduce

```
v0.10.0 20b1486d11
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D select array_value(1,2,3);
┌──────────────────────┐
│ array_value(1, 2, 3) │
│      integer[3]      │
├──────────────────────┤
│ [1, 2, 3]            │
└──────────────────────┘
D create table test as select array_value(1,2,3);
D describe test;
┌──────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│     column_name      │ column_type │  null   │   key   │ default │  extra  │
│       varchar        │   varchar   │ varchar │ varchar │ varchar │ varchar │
├──────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ array_value(1, 2, 3) │ INTEGER[3]  │ YES     │         │         │         │
└──────────────────────┴─────────────┴─────────┴─────────┴─────────┴─────────┘
D copy test to 'array.parquet';
Error: Not implemented Error: Unimplemented type for Parquet "INTEGER[3]"
```

### OS:

Windows 10

### DuckDB Version:

0.10.0

### DuckDB Client:

CLI

### Full Name:

Stefano Lanzavecchia

### Affiliation:

private

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### LINKED ISSUE #10667: The C-API and ARRAYs
### What happens?

I am using the C-API through an interpreted language and when I use a sequence of calls that looks like this:
```
      row_count←api.duckdb_data_chunk_get_size _data_chunk
      _vector←api.duckdb_data_chunk_get_vector _data_chunk data_chunk_ind
      _column_type←api.duckdb_vector_get_column_type _vector
      type_id←api.duckdb_get_type_id _column_type
 ```
 `type_id` is assigned 0 (DUCKDB_TYPE_INVALID). Indeed (?) `duckdb_type`'s enum from [`dubkdb.h`](https://github.com/duckdb/duckdb/blob/82e85721e40d13d1cd0571c92674c28e9a5c90c5/src/include/duckdb.h#L70C1-L136C15) does not contain an entry for ARRAY. Am I missing something or was the C-API not updated to cater for ARRAYs?

You'll have to believe here that the _data_chunk is valid (it was obtained via `duckdb_query -> duckdb_result_get_chunk` from the simple query `select array_value(1,2,3)`.

### To Reproduce

 I will eventually be able to build a full repro in C but it's going to take a while...

### OS:

Windows

### DuckDB Version:

0.10.0

### DuckDB Client:

C-API

### Full Name:

Stefano Lanzavecchia

### Affiliation:

private

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### LINKED ISSUE #10785: Python connection does not support Array type
### What happens?

When I execute a query with `con = duckdb.connect(...); con.execute(query)` that uses the Array type it fails with:

_duckdb.duckdb.NotImplementedException: Not implemented Error: Unsupported type: "FLOAT[3]"_

note: no relevant extension was found related to the Array type and no extension was loaded into the connection.

### To Reproduce

* run the minimal python script below
* see that the NotImplementedException is raised

Minimal python script that reproduces the error:

```
import duckdb


with duckdb.connect() as con:

    con.execute("CREATE TABLE embeddings (id INTEGER, embedding FLOAT[3]);")
    con.execute("INSERT INTO embeddings VALUES (1, [1.0, 2.0, 3.0]);")

    result = con.execute("SELECT id, embedding FROM embeddings;")
    print(result.fetchall())

```

### OS:

Ubuntu 22.04

### DuckDB Version:

0.10.0

### DuckDB Client:

Python

### Full Name:

Luuk van der Velden

### Affiliation:

Codebeez

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Maxxen**
This PR adds:
- A bunch of different array types `test_all_types()`
- Support for array types/vectors in the C-API
- Support for array import and export in the JSON extension.
- Support for array import and export in the Parquet extension.
- Support for arrays in the Python client, both as python objects and as arrow fixed-size-lists.

This PR does __not__ add
- Support for array types in the Java client
- Support for array types in the Julia client

The tests in these clients based on `test_all_types` have been modified to exclude the new array entries. Im planning to add support for Java in a future PR.

In theory we could also support nested numpy arrays better now now that we have a proper fixed size array type, but this PR makes no changes to how that is handled, nested numpy arrays still become multiple columns.

Since Parquet doesn't have a fixed-size list type, you can't really roundtrip a written array back into DuckDB as you will just get a list type back instead. In the future we may want to annotate array columns with duckdb specific metadata to recognize this during import and perform the conversion (or insert a cast) automagically.

Also fixes some misc array bugs related to flattening.

Closes https://github.com/duckdb/duckdb/issues/10785
Closes https://github.com/duckdb/duckdb/issues/10667
Closes https://github.com/duckdb/duckdb/issues/10663

### CODE REVIEW

**Comment by Tishj** [comment]
Cool, yea about supporting the nested Numpy arrays better, I think this PR makes that step easier to take in the future, right?
<https://github.com/duckdb/duckdb/pull/10826>

**Comment by Mause** [comment]
I've been doing some work on the Julia, Java, and Swift clients. I'll submit that once this is in

File: src/include/duckdb.h [hunk 1]
```diff
@@ -1626,6 +1628,16 @@ Creates a list value from a type and an array of values of length `value_count`
...
 
+/*!
+Creates a array value from a type and an array of values of length `value_count`
+
+* type: The type of the array
+* values: The values for the array
+* value_count: The number of values in the array
+* returns: The value. This must be destroyed with `duckdb_destroy_value`.
+*/
+DUCKDB_API duckdb_value duckdb_create_array_value(duckdb_logical_type type, duckdb_value *values, idx_t value_count);
```
Comment (Mause) [inline src/include/duckdb.h:1639]: Curious why you're passing the value count here, when it's also part of the array type itself?

File: tools/pythonpkg/src/native/python_conversion.cpp [hunk 2]
```diff
@@ -210,6 +210,29 @@ Value TransformListValue(py::handle ele, const LogicalType &target_type = Logica
 	return Value::LIST(element_type, values);
 }
 
+Value TransformArrayValue(py::handle ele, const LogicalType &target_type = LogicalType::UNKNOWN) {
```
Comment (Tishj) [inline tools/pythonpkg/src/native/python_conversion.cpp:213]: The way you can test this is by using `duckdb.execute("query with prepared parameters", {"param_name": param})`

The `target_type` is tested by creating `duckdb.Value(obj, <type object>)`
When scanning the `duckdb.Value` object we unwrap it and pass the enclosed type object as `target_type`

File: src/include/duckdb.h (hunk 1 above)
Comment (Maxxen) [reply src/include/duckdb.h:1639]: No wait, the type we pass in is the child type, not the type of the array itself!

File: extension/parquet/column_writer.cpp [hunk 3]
```diff
@@ -1828,6 +1828,142 @@ void ListColumnWriter::FinalizeWrite(ColumnWriterState &state_p) {
...
+	auto &state = state_p.Cast<ArrayColumnWriterState>();
+
+	auto array_size = ArrayType::GetSize(vector.GetType());
+	auto &validity = FlatVector::Validity(vector);
+
+	// write definition levels and repeats
+	idx_t start = 0;
+	idx_t vcount = parent ? parent->definition_levels.size() - state.parent_index : count;
+	idx_t vector_index = 0;
+	for (idx_t i = start; i < vcount; i++) {
```
Comment (Mytherin) [inline extension/parquet/column_writer.cpp:1901]: This is rather complex code and it looks like it's mostly a copy-paste from the `ListColumnWriter::Prepare` - could we perhaps unify the code (or just cast arrays to lists for parquet writing? we have a `CastColumnReader` for reading, we could similarly consider creating a `CastColumnWriter` for writing)

File: extension/parquet/column_writer.cpp [hunk 4]
```diff
@@ -1916,6 +2052,40 @@ unique_ptr<ColumnWriter> ColumnWriter::CreateWriterRecursive(vector<duckdb_parqu
...
+		repeated_element.repetition_type = FieldRepetitionType::REPEATED;
+		repeated_element.num_children = 1;
+		repeated_element.__isset.num_children = true;
+		repeated_element.__isset.type = false;
+		repeated_element.__isset.repetition_type = true;
+		repeated_element.name = "array";
+		schemas.push_back(std::move(repeated_element));
+		schema_path.emplace_back("array");
+
+		auto child_writer = CreateWriterRecursive(schemas, writer, child_type, "element", schema_path, child_field_ids,
```
Comment (Mytherin) [inline extension/parquet/column_writer.cpp:2084]: Same here - it looks like this is identical to the list code above

File: tools/pythonpkg/src/numpy/array_wrapper.cpp [hunk 5]
```diff
@@ -285,6 +285,22 @@ struct ListConvert {
 	}
 };
 
+struct ArrayConvert {
+	static py::tuple ConvertValue(Vector &input, idx_t chunk_offset, const ClientProperties &client_properties) {
+		auto val = input.GetValue(chunk_offset);
```
Comment (Mytherin) [inline tools/pythonpkg/src/numpy/array_wrapper.cpp:290]: In light of https://github.com/duckdb/duckdb/pull/10826 reworking `ListConvert` - this should ideally also just call/be unified with `ListConvert`, but perhaps @Tishj can pick that up in his rework

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some comments:

File: extension/parquet/column_writer.cpp (hunk 3 above)
Comment (Maxxen) [reply extension/parquet/column_writer.cpp:1901]: The issue with unifying with the list is that we 1. don't have a list_entry_t child vector and 2. need to write validity differently. I think casting would be a good idea but I wasn't sure where to perform the cast. Ill have a look at the `CastColumnReader` and see if I can do something similar for the write-side.

Comment (Maxxen) [reply extension/parquet/column_writer.cpp:1901]: We can't template virtual functions so we'd have to template the whole class

Comment (Maxxen) [reply extension/parquet/column_writer.cpp:1901]: How about I just subclass the ListColumnWriter and override Analyze, Prepare, Write?

**Comment by Maxxen** [comment]
@Mytherin CI failure looks unrelated

**Pull request merged**